framework/src/modules/interoperability/mainchain/commands/submit_mainchain_cross_chain_update.ts

```diff
@@ -237,23 +184,116 @@ export class SubmitMainchainCrossChainUpdateCommand extends BaseCrossChainUpdate
 		} catch (error) {
 			context.eventQueue.restoreSnapshot(eventSnapshotID);
 			context.stateStore.restoreSnapshot(stateSnapshotID);
+
 			logger.info(
 				{ err: error as Error, moduleName: ccm.module, commandName: ccm.crossChainCommand },
 				'Fail to execute beforeCrossChainMessageForwarding.',
 			);
+
 			await this.internalMethod.terminateChainInternal(context, ccm.sendingChainID);
+
 			this.events.get(CcmProcessedEvent).log(context, ccm.sendingChainID, ccm.receivingChainID, {
 				code: CCMProcessedCode.INVALID_CCM_BEFORE_CCC_FORWARDING_EXCEPTION,
 				result: CCMProcessedResult.DISCARDED,
 				ccm,
 			});
+
+			return false;
+		}
+
+		return true;
+	}
+
+	// https://github.com/LiskHQ/lips/blob/main/proposals/lip-0049.md#forward
+	private async _forward(context: CrossChainMessageContext): Promise<void> {
+		const { ccm } = context;
+		const { ccmID, encodedCCM } = getEncodedCCMAndID(ccm);
+
+		if (!(await this.verifyCCM(context, ccmID))) {
+			return;
+		}
+
+		// create a state snapshot
+		const stateSnapshotID = context.stateStore.createSnapshot();
+		const eventSnapshotID = context.eventQueue.createSnapshot();
+
+		let ccmFailed = false;
+
+		const chainAccountStore = this.stores.get(ChainAccountStore);
+		const receivingChainAccount = await chainAccountStore.getOrUndefined(
+			context,
+			ccm.receivingChainID,
+		);
+
+		// do not continue, if
+		// the chain account does not exist
+		// the chain has REGISTERED status
+		if (!receivingChainAccount || receivingChainAccount.status === ChainStatus.REGISTERED) {
+			ccmFailed = true;
+		} else {
+			// https://github.com/LiskHQ/lips/blob/main/proposals/lip-0043.md#sidechain-registration-process
+			// Liveness condition is only applicable to ACTIVE status chains
+			const live = await this.internalMethod.isLive(
+				context,
+				ccm.receivingChainID,
+				context.header.timestamp,
+			);
+			if (!live) {
+				ccmFailed = true;
+
+				/**
+				 * If the receiving chain is active, it means it violated the liveness condition, and we terminate it.
+				 * If the receiving chain is already terminated, terminateChain does nothing.
+				 */
+				await this.internalMethod.terminateChainInternal(context, ccm.receivingChainID);
+
+				/**
+				 * A sidechain terminated message is returned to the sending chain
+				 * to inform them that the receiving chain is terminated.
+				 */
+				await this.internalMethod.sendInternal(
+					context,
+					EMPTY_FEE_ADDRESS,
+					MODULE_NAME_INTEROPERABILITY,
+					CROSS_CHAIN_COMMAND_SIDECHAIN_TERMINATED,
+					ccm.sendingChainID,
+					BigInt(0),
+					CCMStatusCode.OK,
+					codec.encode(sidechainTerminatedCCMParamsSchema, {
+						chainID: ccm.receivingChainID,
+						stateRoot: receivingChainAccount?.lastCertificate.stateRoot,
+					}),
+				);
+			}
+		}
+
```

Comment:
I think we need to take a snapshot of the current state (and perhaps events) here.